core/controllers/classifier.py

```diff
@@ -0,0 +1,92 @@
+# Copyright 2017 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Controllers for communicating with the VM for training classifiers."""
+
+import hashlib
+import hmac
+import json
+
+from core.controllers import base
+from core.domain import classifier_domain
+from core.domain import classifier_services
+from core.domain import config_domain
+
+import feconf
+
+
+def validate_request(handler):
+    """Decorator that checks if the incoming request from the VM is valid."""
+    def test_is_valid(self, **kwargs):
+        message = self.request.get('payload')
+        message = json.loads(message)
+        signature = message.get('signature')
```

Comment:
Just wondering, do you think it's possible to do self.request.get('signature') and self.request.get('vm_id'), so that the signature is a different thing from the payload and doesn't need to be popped? We introduce a 'payload' key for requests coming in from the client just to keep things tidy, but the signature here is conceptually separate so it would be nice to have it as a separate field if possible.

Alternatively, if you want to go with self.payload, then you could have the payload contain three keys: message, signature, and vm_id. That might be cleaner and more consistent with other requests. You also don't really need a decorator to guard the handler; you could just have a regular function called `verify_signature(message, vm_id, received_signature)` that you call from within the handler.

(I know this is a bit different from what was originally suggested, sorry about that.)